MultiFreeText Questions, new Stack and Queue generator, new Huffman Coding input2 option

Fabianstw opened the pull request:
[Test location](https://tcs.uni-frankfurt.de/algo-learn-testing/pull_request_21/en)

In this PR are following new features included:
- MultiFreeTextQuestions:
  - questions with more than one input field
  - different checkFormat feedback variations
- DrawList & DrawTable:
  - new display option with md DrawList
  - more extra feature options for DrawTable
- Stack and Queue generator:
  - only input variant | **more variations in the future**
  - tests for Stack and Queue Class
  - tests like example.test.ts for Stack and Queue generator 
- Huffman Coding new variation input2:
  - ask user for a specific coding of each char 
  - tests for huffmanCoding generator

holgerdell commented:
@Fabianstw
In this pull request, you used "prop drilling" in the markdown renderer, and you pushed FreeTextAnswer etc. data down to the input fields. I'd like to avoid this using Reacts concept of "context", and `Mardown.tsx` should not depend on any specific Question type.

In 5bafee7fd715ec9afc936f8aecbfe1e231d70530 on `main`, I added a somewhat generic `useFormField(key: string)` hook and a `FormInputField` component which can be called from `Markdown.tsx`. If you agree with this solution, you need to prepare a form context and the necessary data before calling <Markdown> like so:
```javascript
const value: TextFieldState = ... // prepare a value of type TextFieldState
<formContext.Provider value={...}><Markdown ...></formContext.Provider>
```

Fabianstw commented:
> @Fabianstw In this pull request, you used "prop drilling" in the markdown renderer, and you pushed FreeTextAnswer etc. data down to the input fields. I'd like to avoid this using Reacts concept of "context", and `Mardown.tsx` should not depend on any specific Question type.
> 
> In [5bafee7](https://github.com/goethe-tcs/algo-learn/commit/5bafee7fd715ec9afc936f8aecbfe1e231d70530) on `main`, I added a somewhat generic `useFormField(key: string)` hook and a `FormInputField` component which can be called from `Markdown.tsx`. If you agree with this solution, you need to prepare a form context (which required extracting all requested input field information) and the necessary data before calling `<Markdown>` like so:
> 
> ```js
> const value: TextFieldState = ... // prepare a value of type TextFieldState
> <formContext.Provider value={...}><Markdown ...></formContext.Provider>
> ```

I removed the prop drilling and replaced it with your provided useContext. I made some changes to `TextFieldState` and removed `FormInputField` since most of its functionalities are already in `CustomInput` (mostly the react styling).

holgerdell commented:
Übersetzung "result" etc. fehlt

Fabianstw commented:
> Übersetzung "result" etc. fehlt

Übersetzungen sollte alle funktionieren (bis auf die eben besprochenen). 

Für die Funktionen `getSize` und `amountElements`, habe ich unter die Frage zwei kleine Erklärungen geschrieben. Zwar auch erstmal nur eine kurzzeitige Lösung, aber dadurch weiß man was die Funktionen machen. 

Testet es gerne nochmal 👍

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+    )
+  }
+
+  return <>{inputElement}</>
```
kann vereinfacht werden zu `return inputElement`

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+              ref={first ? firstInputRef : null}
+              key={`newline-input-${inputID}`}
+              autoFocus
+              disabled={questionMode === "correct" || questionMode === "incorrect"}
```
this should have been computed by the context provider, not here.

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+              onChange={(e) => {
+                setText ? setText(e.target.value) : ""
+              }}
+              onFocus={() => setIsInputFocused(true)}
```
what is this needed for?

Fabianstw replied:
if the field is selected, we can display formatFeedback otherwise only a red border. 

So feedback about the format will only be displayed if the user is inside the textfield

holgerdell replied:
makes sense, thank you

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+
+  const inputBorderColor = modeID === "invalid" ? "border-red-500 focus:border-red-500" : ""
+
+  const alignDiv: string = "w-full h-full"
```
remove the variable and hard-code it below.

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+import { useFormContext } from "@/hooks/useFormContext.ts"
+import { useTranslation } from "../hooks/useTranslation"
+
+export const CustomInput: React.FC<{ id: string }> = ({ id }: { id: string }) => {
```
rename this to FormInputField and move it to its own file under `@/components/ui/`. The component should be UI-only and not perform any logic from the Question (I think the only such logic left is disabled and invalid)

holgerdell commented on front-end/src/components/Markdown.tsx:
```diff
@@ -102,6 +104,9 @@ export const MarkdownTreeNode: FunctionComponent<{
     return <DrawTable table={parseTreeNode.child} />
   }
+  if (parseTreeNode.kind === "list") {
+    return <DrawList list={parseTreeNode.child} />
```
why is this needed? Lists can be included using standard markdown syntax:
```
- item a
- item b
```

Fabianstw replied:
It can be nested List and instead of `-` it uses a list point

holgerdell replied:
Nested lists should also use standard markdown syntax:
```
- item a
  - subitem 1
  - subitem 2
- item b
```

I don't think we need to allow the question generator to change the style of the list. If you prefer list points in general, change this globally.

Fabianstw replied:
But we currently don't have lists in Markdown. This would be just rendered as normal Markdown text and the spaces in front of `subitem 1` and line breaks would be removed. 

With this we could create a list block. Space to the top and bottom and always the same style  

Or am I missing something?

holgerdell replied:
- My point is that there is standard syntax for lists in Markdown, which we should use instead of a `list` code-block. Yes, this requires finding the correct regex etc.
- We are slowly reaching a level of complexity for our parser where we may just want to use a standard markdown parser such as https://marked.js.org instead of rolling our own. Probably @AbcAeffchen will object here?
- If you want, you can decide to keep the quick solution using a code block, and put the proper list syntax to the someday/maybe todo list.

Fabianstw replied:
Yes I understand what you mean. 

https://www.markdownguide.org/basic-syntax/#unordered-lists
There are 3 different types for unordered Markdown lists. I took the list with the `-`, maybe not the same regex but mainly it should do the same. It parses the list-string line by line into a string tree.

Fabianstw replied:
I would like to keep the quick solution and add an issue once this PR has been merged, so we don't forget it

holgerdell commented on front-end/src/locales/de.json:
```diff
@@ -105,5 +105,6 @@
   "train.skills.desc": "Sammle Erfahrungspunkte (XP) durch Übung! Wähle ein Thema:",
-  "your-learning-progress": "Dein Lernfortschritt"
+  "your-learning-progress": "Dein Lernfortschritt",
+  "provideFeedbackCheckFormat": "Hie werden wir dir Feedback zum Format deiner Antwort geben"
```
Hie -> Hier

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+      <div>
+        {spacing}
+        <div className="flex flex-row items-center">
+          <div className="mr-2">{promptElement}</div>
```
Der Prompt sollte meines Erachtens einfach im Markdown-Text stehen, damit auch Lückentexte etc. möglich sind. Wenn man es in der Tabelle haben will, nutzt man Tabellen-Notation:

| --- | --- |
| Prompt | Eingabfeld |
| Fragentext1 | {{! inputfeld1#bla#blubb !}} |
| Fragentext2 | {{! inputfeld2#bla#blubb !}} |

Dadurch wird der Prompt in Markdown gerendert. Nur das Eingabefeld muss hier in aktuell CustomInput erzeugt werden, nicht der Prompt. Oder übersehe ich etwas?

Fabianstw replied:
Ich weiß was du meinst und ich hatte es auch schon anders implementiert. 

Die Darstellung in einer Tabelle funktioniert (nutze ich aktuell auch bei der Stack und Queue Frage). 
Ich würde den Prompt weiterhin so lassen wie er aktuell ist, da sich dadurch das Feedback, das InputFeld und der Prompt besser aneinander ausrichten (diese sind dann ein zusammenhängender Block). 

Lückentexte werden noch nicht unterstützt. Ich habe es nicht hinbekommen Feedback darunter anzuzeigen, welches zentriert unter dem Feld ist, aber dennoch größer ist als das Feld. Da mir das nicht gefallen hat habe ich es rausgenommen.

Fabianstw replied:
Soll ich das noch einfügen oder wollen wir das einfügen, wenn uns Bedarf einfällt?

holgerdell commented on front-end/src/components/CustomInput.tsx:
```diff
@@ -0,0 +1,162 @@
+            />
+            {isInputFocused && text && feedback && (
+              <FeedbackComponent
+                formatFeedback={feedback ? feedback : ""}
```
kann zu `feedback ?? ""` verkürzt werden.

Fabianstw replied:
I check if `text` is not empty, because otherwise the user still has to type something. It's not necessary, because the generator should return an empty string if the text is empty. I'll take it out. 

But I would keep `isInputFocused`

holgerdell reviewed (commented):
Hier ist eine erste Runde Feedback auf den Code, hauptsächlich auf die MultiInputText-Geschichte.

Fabianstw commented on front-end/src/hooks/useFormContext.ts:
```diff
@@ -1,16 +1,20 @@
-  placeholder?: string // the placeholder text of the input (usually in gray)
-  invalid?: boolean // true if the current value cannot be submitted (e.g. syntactically invalid)
+  placeholder: string // the placeholder text of the input (usually in gray)
+  modeID?: MODE // the mode of the input field
```
I changed it, but I didn't add `width`, because "Lückentexte" are currently not supported

holgerdell commented on front-end/src/components/ui/FormInputField.tsx:
```diff
@@ -1,29 +1,27 @@
 
-export const CustomInput: React.FC<{ id: string }> = ({ id }: { id: string }) => {
-  const useInputContext = useFormContext()
+export const FormInputField: React.FC<{ id: string }> = ({ id }: { id: string }) => {
```
zweite Typannotation

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -41,30 +41,39 @@ const translations: Translations = {
       " Also choose as the left node, the node with the smaller weight.",
+    multiInputText: `Suppose we have the following table, which represents how often each character appears in a given string:
+{{0}}
+What could be a correct **Huffman-Coding** for each character?
```
Besser als Aufforderung formulieren: `Construct a correct Huffman-Code for the given string.`

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -41,30 +41,39 @@ const translations: Translations = {
     name: "Berechne eine Hufmann-Codierung",
     description: "Bestimme die Huffman-Codierung eines gegebenen Strings",
-    text: 'Sei "*{{0}}*" ein {{1}}. Was ist eine korrekte **Huffman-Codierung** für diesen {{1}}?',
+    text: 'Sei "*{{0}}*" ein {{1}}. Was ist eine korrekte **Huffman-Codierung** für diesen {{1}}? Bitte ignorieren die Leerzeichen.',
```
Grammatik: `Bitte ignorieren die Leerzeichen.`

holgerdell commented on shared/src/question-generators/StackQueue/QueueGenerator.ts:
```diff
@@ -0,0 +1,818 @@
+
+const translations: Translations = {
+  en: {
+    name: "Queue-Implementation using an Array",
```
`Queues implemented as fixed arrays` (im Gegensatz zu dynamischen Arrays)

holgerdell replied:
wobei, solange es sonst keine queue-Aufgabengeneratoren gibt, kann das hier einfach `Queues` heißen

Fabianstw replied:
Habe die Verbesserungen für `Queue` habe auch für `Stack` übernommen.

holgerdell commented on shared/src/question-generators/StackQueue/QueueGenerator.ts:
```diff
@@ -0,0 +1,818 @@
+const translations: Translations = {
+  en: {
+    name: "Queue-Implementation using an Array",
+    description: "Correctly use queue operations",
```
"correctly" kann man wohl streichen. Vielleicht `Perform queue operations`?

holgerdell commented on shared/src/question-generators/StackQueue/QueueGenerator.ts:
```diff
@@ -0,0 +1,818 @@
+    getQueueInfo:
+      "**Note:** The method **getQueue()** returns the complete queue as it is as a string. If the queue is not full, the remaining elements are filled with -1.",
+    toStringInfo:
+      "**Note:** The method **toString()** only returns the part between rear and front. Every other value from the field is ignored.",
```
Die abstrakte Datenstruktur Queue hat keine Operationen `getQueueInfo` oder `toString`, wir sollten uns strikt an die abstrakte Datenstruktur halten (wie zum Beispiel in CLRS definiert). Ich würde diese Operationen vollständig streichen bzw. die Frage entsprechend anders formulieren.

Fabianstw replied:
Ich verstehe was du meinst. Die Methoden `.getQueue()` und `.toString()` existieren eigentlich nicht. Ich würde gerne aber eine Frage haben, bei welcher der Nutzer angeben soll, wie das Array der Queue aussieht. 

Ich würde ungerne die Frage streichen, sondern lieber anders formulieren. Wäre `.getArray()` oder `getField()` besser? Dann würde ich `.toString()` streichen und `.getQueue()` umbenennen. 

So auch bei StackQuestion

Fabianstw replied:
Wären dies bei der Queue dann nur die Operationen `enqueue` und `dequeue`?

holgerdell replied:
Auch `isEmpty`, aber sonst nichts, siehe Folie 6: https://files.tcs.uni-frankfurt.de/algo1/introtodatastructures.pdf
Wir müssen uns strikt an die Operationen der abstrakten Datenstruktur halten, das ist genau das Lernziel für diese Aufgabe, mit nur diesen Operationen umzugehen, sowie die interne Darstellung zu verstehen.

Die Frage "wie sieht das Array aus" ist sinnvoll, jedoch darf dies nicht über die Operationen der Datenstruktur geschehen. Stattdessen "Gib den Zustand der Datenstruktur an" mit Beispiel, wie die Eingabe formatiert sein soll.

Jetzt oder in einem zukünftigen pull request: Kannst du außerdem die Aufgaben 4.1a, 4.1c und 4.2 in genau der Formulierung von dem Übungsblatt als Variante implementieren? https://goethe-tcs.github.io/algo1-exercises/ALGO1-Blatt-04-introtodatastructures.pdf

Fabianstw replied:
Okay, dann werde ich dies verändern, indem ich die Frage nach dem State des Feldes als weiteren Absatz einfüge. Also außerhalb der aktuellen Tabellenform.

Fabianstw replied:
Die weiteren Aufgaben schreibe ich mir auf meine Todo-Liste und ergänze diese in einem zukünftigen PR. Dann wird es in diesem PR nicht noch größer.

holgerdell commented on shared/src/question-generators/StackQueue/QueueGenerator.ts:
```diff
@@ -0,0 +1,818 @@
+    queueEmpty: "Currently the queue is empty.",
+    queueContainsValues: `The queue currently contains the following elements:`,
+    multipleChoiceText:
+      `Consider having a **Queue "{{0}}"**, which can store at most` +
```
Grammatik ("having"), und hier müsste was über das Array stehen.
`Consider a **Queue "{{0}}"** implemented as a fixed array of size ...`

Fabianstw replied:
Habe dies bei Stacks auch geändert zu `Consider a **Stack "{{1}}"** implemented as a dynamic/fixed array of size ...`

holgerdell reviewed (commented):
Ich habe wir mal die Texte angesehen und kleine Kommentare dagelassen

holgerdell commented on shared/src/question-generators/StackQueue/QueueGenerator.ts:
```diff
@@ -50,13 +50,13 @@ const translations: Translations = {
-      `Angenommen du hast eine **Queue "{{0}}"**, welche maximal` +
-      " {{1}} " +
-      `Elemente speichern kann. {{2}} {{3}} **Welche Aussagen können wir nun über die Queue treffen?**`,
+      `Angnommen du hast eine **Queue "{{0}}"** implementiert als fixer Array der Größe` +
```
- `Angnommen`
- `fixer` -> `festes`

holgerdell commented on shared/src/question-generators/StackQueue/QueueGenerator.ts:
```diff
@@ -50,13 +50,13 @@ const translations: Translations = {
-      `Angenommen du hast eine **Queue "{{0}}"**, welche maximal` +
-      " ${{1}}$ " +
-      `Elemente speichern kann. {{2}} **Wir führen die folgenden Operationen aus:** {{3}}
+      `Angnommen du hast eine **Queue "{{0}}"** implementiert als fixer Array der Größe` +
```
hier auch

holgerdell commented on shared/src/question-generators/StackQueue/StackGenerator.ts:
```diff
@@ -13,53 +13,51 @@ import { t, tFunction, tFunctional, Translations } from "@shared/utils/translati
-    name: "Stack-Implementation using an Array",
-    description: "Correctly use stack operations",
+    name: "Stack",
+    description: "Perform queue operations",
```
stack

holgerdell commented on shared/src/question-generators/StackQueue/StackGenerator.ts:
```diff
@@ -13,53 +13,51 @@ import { t, tFunction, tFunctional, Translations } from "@shared/utils/translati
-    dynamic: "dynamisch",
-    notDynamic: "nicht dynamisch",
+    dynamic: "dynamischen",
+    notDynamic: "fixen",
```
dynamisches
festes

holgerdell commented on shared/src/question-generators/StackQueue/StackGenerator.ts:
```diff
@@ -13,53 +13,51 @@ import { t, tFunction, tFunctional, Translations } from "@shared/utils/translati
-      `Angenommen du hast einen **{{0}} Stack "{{1}}"**, welcher maximal` +
-      " {{2}} " +
-      `Elemente speichern kann. {{3}} {{4}} 
+      `Angenommen du hast einen **Stack "{{1}}"** implementiert als {{0}} Array der Größe` +
```
Hier und sonstwo: `Angenommen du hast` ist umgangsprachlich. Besser: `Betrachte einen`

holgerdell commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -39,40 +39,40 @@ import { t, tFunction, tFunctional, Translations } from "@shared/utils/translati
     textTable: `Angenommen wir habe die folgende Tabelle, welche angebibt, wie oft jeder Buchstabe in einem gegebenen String vorkommt:
 {{0}}
-        Was wäre eine korrekte **Huffman-Codierung** für jeden Buchstaben?`,
+        Was wäre ein korrekter **Huffman-Code**?`,
```
Wie im englischen als Aufforderung umformulieren: `Konstruiere einen korrekten`

holgerdell replied:
betrifft auch andere Texte in dieser Aufgabe

Fabianstw replied:
Sollte alle Aufgabenstellung jetzt direkt formuliert haben.

holgerdell reviewed (commented):
weitere sprachliche Kommentare

Fabianstw commented:
Ich habe die Option für `choice` sowohl bei Stack als auch bei Queue erstmal rausgenommen. Ich finde die Variation ist nicht gut genug und passt aktuell nicht. Den Code habe ich mir aber zwischengespeichert. 

Es ist jetzt nur noch die `input` Variation vorhanden, auch nur noch mit den Optionen `push` / `pop` / `isEmpty` (so auch bei der Queue). Weitere Variationen würde ich ergänzen wenn dies gemergt ist. 

Der PR sollte hierdurch auch übersichtlicher geworden sein 👍

holgerdell commented:
Ich habe mir nur die Seite angeschaut und die sieht gut aus!

Nur eins gefällt mir noch nicht, bitte sprachspezifische Ausdrücke entfernen, also `(void function)` entweder durch "none" oder mir sogar noch lieber den leeren String "" ersetzen.

Hier wird das Array übrigens mit Index und Value angezeigt, was mir gefällt. Vielleicht sollte dieser Stil in den anderen pull-requests für die Anzeige von Arrays auch benutzt werden?

holgerdell commented:
Abgesehen davon kannst du es aber von meiner Seite squashen und mergen

AbcAeffchen commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -342,25 +340,45 @@ export const huffmanCoding: QuestionGenerator = {
           return {
             correct: true,
-            message: "t(feedback.correct)",
+            message: tFunction(translations, lang).t("feedback.correct"),
```
`t(translation, lang, "feedback.correct")`

AbcAeffchen commented on shared/src/question-generators/huffman-coding/huffmanCoding.ts:
```diff
@@ -342,25 +340,45 @@ export const huffmanCoding: QuestionGenerator = {
         return {
           correct: false,
-          message: "t(feedback.incomplete)",
+          message: tFunction(translations, lang).t("feedback.incomplete"),
```
`t(translation, lang, "feedback.incomplete")`

AbcAeffchen reviewed (commented):
I finally found some time to look into this. I only found some minor issues.
- Queue and Stack seem to be implemented in a quite complicated way. This is often the way to do it if you need the performance e.g. in C++. However this is not the case here. We only have tiny toy examples. So the code could be much simpler and better understandable.
- In the queue question I assumed the front to be the right end of the array and the back to be the left side, but it is the other way around. For the stack the top is at the right side. So maybe there should be an indication for that in both cases.
- When I open the stack or queue question in a small window, it scrolls to the bottom. It doesn't do this for other questions like oh-notation, where the question would also be long enough that there it could be scrolled down. maybe this has something to do with the multi input question? However, when the page opens it should show the top so I can read the question.
- There are also some comments in the code.

I skipped the huffman stuff since there is an upcomming merge from main.

Fabianstw commented on shared/src/question-generators/StackQueue/Queue.ts:
```diff
@@ -0,0 +1,219 @@
+/**
+ * This class represents a queue
+ */
+export class Queue {
```
I removed the minFind() function, because it is unused. And I add option to use it with any value.

Fabianstw commented on shared/src/question-generators/StackQueue/Queue.ts:
```diff
@@ -0,0 +1,219 @@
+  private numberOfElements: number
+  constructor(size: number) {
+    this.size = size
+    this.queue = new Array(size).fill(0) as number[]
```
Simplified it as discussed 👍

Fabianstw commented on front-end/src/components/DrawTable.tsx:
```diff
@@ -1,29 +1,116 @@
  * A component that returns a table
  * @param table The table to be drawn (passed as md format)
+ * @param setText
+ * @param state
```
No, I updated the JSdoc
